### PR TITLE
feat(core): validate maximum amount of resources in a stack

### LINKED ISSUE #276: Warn/fail synthesis when reached the 200 resource limit
Stack synthesis should fail if number of resources exceeds 200 (potentially warn if we are at 80% of that).

### PR DESCRIPTION
**Pull request by rzamana**
closes #276
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by rzamana** [comment]
The PR Linter is failing because it doesn't have a README for this feature! Where this should be? 
In the README for the `cx-api`? 
Or this is not a feature, but instead a fix?

File: packages/@aws-cdk/core/lib/stack.ts [hunk 1]
```diff
@@ -753,6 +768,17 @@ export class Stack extends CoreConstruct implements ITaggable {
...
     const outPath = path.join(builder.outdir, this.templateFile);
+
+    if (this.node.tryGetContext(cxapi.VALIDATE_STACK_RESOURCE_LIMIT)) {
+      const resources = template.Resources || {};
+      const numberOfResources = Object.keys(resources).length;
+
+      if (numberOfResources > this.maxResources) {
+        throw new Error(`Number of resources: ${numberOfResources} is greater than allowed maximum of ${this.maxResources}`);
+      } else if (numberOfResources >= (this.maxResources * 0.8)) {
+        Annotations.of(this).addWarning(`Number of resources: ${numberOfResources} is approaching allowed maximum of ${this.maxResources}`);
```
Comment (rix0rrr) [inline packages/@aws-cdk/core/lib/stack.ts:779]: Please change this into `addMessage`.

A `warning` may have the effect of breaking deployment, if deployment is run using `--strict` (which turns warnings into errors).

This is a warning, but it's a different kind of warning from the regular ones.

**Review by rix0rrr — DISMISSED** [review]
I like it, thanks!

With one change: I think we don't need the feature flag, this feels like something that can be enabled wholesale. The limit is unlikely to change anytime soon (it recently changed), and it's unlikely that there are valid use cases for having >500 resources in an undeployable template.

If you feel uncomfortable doing this, then I'd suggest putting the resource limit in a context settings, so people could set something like:

```
"context": {
  "@aws-cdk/core.stackResourceLimit": 1000,
}
```

If they feel like it (and potentially we could have have `0` mean "no limit" to switch it off).

But honestly I don't think there's a need for such an elaborate mechanism.

File: packages/@aws-cdk/core/lib/stack.ts (hunk 1 above)
Comment (rzamana) [reply packages/@aws-cdk/core/lib/stack.ts:779]: Would you mean `addInfo`?

Because the `addMessage` is a private method from `Annotations`!

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/stack.ts]: Oh, doh.

**Review by rix0rrr — DISMISSED** [review]

**Review by rix0rrr — APPROVED** [review]

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**